### ISSUE TITLE
COM Port Not Appearing/Custom Firmware Issue

### ISSUE THREAD

**Issue by 1iggy2** [ArduPilot/MissionPlanner#2103]
#### Issue details
_Please describe the problem, or desired feature_
I was having difficulty finding the Install Firmware page on my old install. I was attempting to install a "Custom" version of ArduPilot. (In this case my own Ubuntu/waf build of ArduPlane). Since I couldn't go through the Install Firmware page, then the Load Custom Firmware option I decided to reinstall Mission Planner. I reinstalled three ways, first from the Mission Planner Latest Installer, then through the github link to the most recent version, currently from this page selecting "MissionPlanner-1.3.63.msi" each time I run into the same issue, When I plug in the device the COM port, in my case COM4, flashes as available and then disappears. I saw that this is commonly a driver issue and I made sure all my driver were up to date through the device manager. Now I am unable to connect to my Pixhawk4 to flash my fake custom firmware to test my ability to modify the ArduPilot source code. 

Additional Note: The initial install of Mission Planner on my machine was from the Microsoft Store. After each iteration of an attempt to fix my issue the previous version of Mission Planner was uninstalled using the Windows Add or Remove Programs tool. I did attempt to install my "custom" arduplane software through QGroundControl at one point flashing a .bin file to it, Since then I had attempted to flash the .APJ file from the new Mission Planners. I have no idea if they were successful, But the software seemed to carry out the flash nominally. I just couldn't connect or see the COM port after, but I never tried to connect directly before flashing.

#### Version
_What version was the issue encountered with_
Holybro Pixhawk4
ASUS Windows 10 Laptop
Mission Planner 1.3.63 built 1.3.7014.32511
Arduplane, Custom compiled firmware (I kept full notes for my project team and I will provide them below)

#### Platform
[  ] All
[  ] AntennaTracker
[  ] Copter
[X] Plane
[  ] Rover
[  ] Sub

#### Airframe type
_What type of airframe (flying wing, glider, hex, Y6, octa etc)_
Quadplane Octo Xish
I am attempting to customize this eventually to allow for an Octo "I" configuration. See my other issue report for more information (https://github.com/ArduPilot/ardupilot/issues/10809). If you are so inclined to help with the issues I commented at the end of the post I would GREATLY appreciate it.

#### Hardware type
_What autopilot hardware was used? (pixhawk, pixracer, PX4FMU etc)_
Holybro Pixhawk4

#### Logs
_Please provide a link to any relevant logs that show the issue_
![Missionpannerissue](https://user-images.githubusercontent.com/46946416/54662663-1f872f80-4ab5-11e9-83b1-7d87a24d9b33.JPG)
This is an image of my screen when I select the COM4 port when it appears and hit connect before it disappears, When COM4 appears I get the upbeat musical tone of connection and then when it disappears I get the down musical tone of disconnection.

As promised above here is every single step I took when building the firmware: (Let me know if you see an error)

Install Ubuntu on Windows 10
Be in Windows 10 64-bit version (Default)
Go to: Control Panel>Programs>Turn Windows Features On Or Off
Enable “Windows Subsystem for Linux” then hit “OK”
Restart Computer (Not necessary if Windows Subsystem for Linus already enabled)
Open Microsoft Store
Search “Ubuntu”
Install Ubuntu
Open Ubuntu after downloaded and allow to install
Set your UNIX username and password, and remember them

Setting up the waf Build Environment on Windows 10 using WSL
Run “sudo apt-get update”
Enter your password
Run “sudo apt-get upgrade“
Enter “Y” to continue, Allow to install (Took approx 7 minutes)
Enter “git clone https://github.com/ardupilot/ardupilot.git“ 
(may need to change this address to our own address once we have customized code)
Enter “cd ardupilot”
Enter “git submodule init”
Enter “git submodule update --recursive“ (Approx 2 minutes)
Enter “./Tools/scripts/install-prereqs-ubuntu.sh”
Enter “Y” (Approx 25 Minutes)
Enter your password when prompted and let it continue to run.
Enter “y” when prompted
Enter “y” when prompted
Allow to finish (Approx 1 minute)
Enter “nano ~/.profile”
Delete line “export PATH=/opt/gcc-arm-none-eabi-6-2017-q2-update/bin:$PATH”
Hit “Ctrl + O”
Hit “Enter”
Hit “Ctrl + X”
Enter “sudo apt-get install gcc-arm-none-eabi”
Enter “Y”
Allow to finish (Approx 5 minutes)
Enter “. ~/.profile“ to reload the .profile script

Building With Waf
Enter “./waf configure --board sitl --debug”
Enter “./waf configure --board Pixhawk4”
Enter “./waf build sub”
Allow to finish (Approx 10 minutes)
Enter File Explorer and navigate to: C:\Users\<username>\AppData\Local\Packages\CanonicalGroupLimited.UbuntuonWindows_79rhkp1fndgsc\LocalState\rootfs\home\<username>\ardupilot\build\Pixhawk4\bin
You may have to go to the View tab and check the box for hidden items to see AppData when you are in <username>
From there you have access to the APJ files, which are assumed to be the firmware to be flashed

**Comment by 1iggy2** [comment]
I reloaded the original firmware and the issue went away. With the "custom" firmware the issue persists

**Comment by magicrub** [comment]
Well, what COM port is it? Which ones are available? The error is from trying to talk to a com Port that isn't there.

**Comment by 1iggy2** [comment]
So the COM Port would connect and disconnect in five seconds. During that time I'd try to connect mission planner to it. That explains the issue of trying to connect to a disconnected COM port. Now when I flash my own build of the base ardupilot github mission planner will try to connect but the MAVlink connection will timeout with no heartbeat detected.

**Comment by 1iggy2** [comment]
This even happens when I build it using Waf using the commands, 
./waf configure --board Pixhawk4 
./waf plane --upload
The screen displays:
![waf build and upload](https://user-images.githubusercontent.com/46946416/54800862-78cc9b80-4c3a-11e9-98a0-dbd37452aca0.JPG)
Then I go into mission planner and select the correct COM port, (which now appears) and make sure the baud rate is at 115200. Then I select CONNECT. It says CONNECTING TO MAVLINK. It tries to connect for 30 seconds and fails with "No Heartbeat Packets Received"
When I select details I get this popup:
![MissionplannerHeartbeaterror](https://user-images.githubusercontent.com/46946416/54800951-e1b41380-4c3a-11e9-8c6c-d8fab24f0d35.JPG)
If I connect the board to QGroundControl it simply says, "Waiting for Vehicle Connection" and never finds the board.

What is the compile_commands.json? How else can I build the firmware so that it works? I will need to modify the source code for my application, and if I cannot flash the board and connect to mission planner I cannot complete my task. Is there a way for someone else to build files that will work and I can simply flash them through the "Load custom firmware" option (This was my original attempt, I would build a .apj and attempt to upload that through both Mission Planner and QGroundControl, but a .bin fil for QGroundControl,  and neither allowed for connection)

**Comment by 1iggy2** [comment]
@magicrub 
Sorry to tag you, but I really need help resolving this issue and I provided some more information.
Thanks!

**Comment by magicrub** [comment]
So it works with stock code but not with your custom code? Well, sounds like your custom code might be involved. Could you share it?

**Comment by magicrub** [comment]
This is probably not a mission planner problem. The couple second connection is the bootloader. After that, the app runs. If the app is dead, I suspect your custom code.

**Comment by 1iggy2** [comment]
So the "custom" code is simply me following the above instructions and I believe cloning the current ArduPilot github files. In no step did I modify the contents of those files. I'll include the arduplane files I have and the bin in my Pixhawk4 folder that was under build.

[ArduPlane.zip](https://github.com/ArduPilot/MissionPlanner/files/2997873/ArduPlane.zip)
[bin.zip](https://github.com/ArduPilot/MissionPlanner/files/2997874/bin.zip)

**Comment by magicrub** [comment]
This is not a bug in Mission Planner or ArduPilot. I suspect it's just a USB driver you're missing from ChibiOS. It's included in Mission Planner, perhaps a reinstall would solve it.

If that doesn't fix it, please post on discuss.ardupilot.org and mention important things like:
- how you built the code (the commands you used),
- computer OS
- version of code (git hash)

**Comment by 1iggy2** [comment]
Hey @magicrub I've uninstalled and reinstalled Mission Planner approx 12 times, I've checked all my drivers for updates and even deleted some that came with Mission Planner to force a reinstall. I've handed off the Pixhawk4 to a teammate so he can work on it on his system. Where can I find the git hash? Would it be sufficient to link to this post to provide my information on the forum as this is a decently complete document of my attempts, including an entire document how I set up my build environment and all the commands I used, also I've provided all my commands I've used and my troubleshooting attempts.
